agent/src/agent/data_extractor/rrd/rrd.py

```diff
@@ -0,0 +1,100 @@
+import os
+import re
+import rrdtool
+
+from typing import List
+from . import repository
+from agent.modules import logger
+from agent.pipeline import Pipeline
+from agent import source
+
+logger_ = logger.get_custom_logger(__name__, os.environ.get('RRD_SOURCE_LOG_FILE_PATH', 'agent_rrd_source.log'))
+
+
+def extract_metrics(pipeline_: Pipeline, start: str, end: str, step: str) -> list:
+    metrics = []
+
+    sources = repository.get_cacti_sources(
+        pipeline_.source.config[source.CactiSource.MYSQL_CONNECTION_STRING],
+        pipeline_.config['exclude_datasources'],
+        pipeline_.config['exclude_hosts']
+    )
+    for cacti_source in sources:
+        base_metric = {
+            'target_type': 'gauge',
+            'properties': _extract_dimensions(cacti_source),
+        }
+        rrd_file_name = _extract_rrd_file_name(cacti_source)
+        if not rrd_file_name:
+            continue
+
+        rrd_file_path = os.path.join(pipeline_.source.config[source.CactiSource.RRD_DIR], rrd_file_name)
+        result = rrdtool.fetch(rrd_file_path, 'AVERAGE', ['-s', start, '-e', end, '-r', step])
+
+        # result[0][2] - is the closest available step to the step provided in the fetch command
+        # if they differ - skip the source as the desired step is not available in it
+        if result[0][2] != int(step):
+            continue
+
+        # contains the timestamp of the first data item
+        data_start = result[0][0]
+        for name_idx, measurement_name in enumerate(result[1]):
+            for row_idx, data in enumerate(result[2]):
+                timestamp = int(data_start) + row_idx * int(step)
+                value = data[name_idx]
+
+                # rrd might return a record for the timestamp earlier then start
+                if timestamp < int(start):
+                    continue
+                # value will be None if it's not available for the chosen consolidation function or timestamp
+                if value is None:
+                    continue
+
+                metric = base_metric.copy()
+                metric['what'] = measurement_name
+                metric['value'] = value
+                metric['timestamp'] = timestamp
+                metrics.append(metric)
+
+    metrics = _add_tags(metrics, pipeline_.tags)
```

Comment:
I thought we agreed to use streamsets stage for this

Reply:
It's necesary also because it adds metadata tags like pipeline_id, pipeline type etc